src/ophyd_async/epics/areadetector/writers/general_hdffile.py

```diff
@@ -0,0 +1,129 @@
+from typing import Any, Iterator, List, Optional
+from urllib.parse import urlunparse
+
+import event_model
+from event_model import (
+    ComposeStreamResource,
+    ComposeStreamResourceBundle,
+    StreamDatum,
+    StreamResource,
+)
+
+from ophyd_async.core import DirectoryInfo
+
+
+@dataclass
+class _HDFDataset:
+    name: str
+    dtype_numpy: Optional[str] = None
+    swmr: bool = False
+    shape: Optional[List[int]] = None
+    multiplier: Optional[int] = 1
+    path: Optional[str] = None
+    device_name: Optional[str] = None
+    block: Optional[str] = None
+    maxshape: tuple[Any, ...] = (None,)
+    dtype: Optional[Any] = None
+    fillvalue: Optional[int] = None
+
+
+SLICE_NAME = "AD_HDF5_SWMR_SLICE"
+
+
+def versiontuple(v):
+    return tuple(map(int, (v.split("."))))
+
+
+class _HDFFile:
+    """
+    :param directory_info: Contains information about how to construct a StreamResource
+    :param full_file_name: Absolute path to the file to be written
+    :param datasets: Datasets to write into the file
+    """
+
+    def __init__(
+        self,
+        directory_info: DirectoryInfo,
+        full_file_name: Path,
+        datasets: List[_HDFDataset],
+        hostname: str = "localhost",
+    ) -> None:
+        self._last_emitted = 0
+        self._hostname = hostname
+
+        if len(datasets) == 0:
+            self._bundles = []
+            return None
+
+        if versiontuple(event_model.__version__) < versiontuple("1.21.0"):
+            path = f"{directory_info.root/full_file_name}"
+            root = str(directory_info.root)
+            bundler_composer = ComposeStreamResource()
+
+            self._bundles: List[ComposeStreamResourceBundle] = [
+                bundler_composer(
+                    spec=SLICE_NAME,
+                    root=root,
+                    resource_path=path,
+                    data_key=ds.name.replace("/", "_"),
+                    resource_kwargs={
+                        "name": ds.name,
+                        "block": ds.block,
+                        "path": ds.path,
+                        "shape": ds.shape,
+                        "multiplier": ds.multiplier,
+                        "timestamps": "/entry/instrument/NDAttributes/NDArrayTimeStamp",
+                    },
+                )
+                for ds in datasets
+            ]
+        else:
+            bundler_composer = ComposeStreamResource()
+
+            uri = urlunparse(
+                (
+                    "file",
+                    self._hostname,
+                    str((directory_info.root / full_file_name).absolute()),
+                    "",
+                    "",
+                    None,
+                )
+            )
+
+            self._bundles: List[ComposeStreamResourceBundle] = [
+                bundler_composer(
+                    mimetype="application/x-hdf5",
+                    uri=uri,
+                    data_key=ds.name.replace("/", "_"),
```

Comment:
Which would make this:
```suggestion
                    data_key=ds.data_key,
```
I would make data_keys with `/` in them an error, not a silent replacement

Reply:
I have a question regarding the removal of `replace("/", "_")` . If I remove this I have to replace `DATA_PATH = /entry/data/data` `DATA_PATH = /entry/data/data` and so on. If not I get a  `regex not matching error` from the bluesky runengine . I tentatively replaced the paths this way in the code but would like to ask for suggestions.

Reply:
They are different things, data_key is the name within the descriptor, and dataset is the path within the HDF file. So for pattern generator I think we want something like:
```python
            _HDFDataset(
                data_key=f"{detector_name}-data,
                dataset=DATA_PATH,
            )
```
I'm not sure quite how you pass the detector name through and into the pattern generator though...

Reply:
In `pattern_generator.py`, PATH parameters were assigned to `name` (now `data_key`). What I meant is that, deleting the `replace("/", "_")` resulted in errors so I had to modify the PATH parameters. The other thing is that, I am not sure I understand passing detector name to `_HDFDataset` since I am not experienced with the usage of this class. Is the device going to assign the parameters in this class? I would be happy to have more context.